klippy/extras/bltouch.py

```diff
@@ -53,54 +59,96 @@ def __init__(self, config):
+        self.next_cmd_time = mcu.clock_to_print_time(cmd_clock)
+        return self.next_cmd_time
+    def verify_state(self, check_start_time, check_end_time, triggered, msg):
+        # Perform endstop check to verify bltouch reports desired state
         prev_positions = [s.get_commanded_position()
                           for s in self.mcu_endstop.get_steppers()]
-        self.mcu_endstop.home_start(home_time, ENDSTOP_SAMPLE_TIME,
-                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME)
+        self.mcu_endstop.home_start(check_start_time, ENDSTOP_SAMPLE_TIME,
+                                    ENDSTOP_SAMPLE_COUNT, ENDSTOP_REST_TIME,
+                                    triggered=triggered)
         try:
-            self.mcu_endstop.home_wait(home_time + MIN_CMD_TIME)
+            self.mcu_endstop.home_wait(check_end_time)
         except self.mcu_endstop.TimeoutError as e:
-            raise homing.EndstopError("BLTouch sensor test failed")
+            raise homing.EndstopError("BLTouch failed to %s" % (msg,))
         for s, pos in zip(self.mcu_endstop.get_steppers(), prev_positions):
             s.set_commanded_position(pos)
+    def raise_probe(self):
+        self.send_cmd('reset')
+        check_start_time = self.send_cmd('pin_up', duration=self.pin_move_time)
+        check_end_time = self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time,
+                          self.pin_up_triggered, "raise probe")
+    def test_sensor(self):
+        if not self.pin_up_touch_triggered or self.pin_up_triggered:
+            # Nothing to test
+            return
+        toolhead = self.printer.lookup_object('toolhead')
+        print_time = toolhead.get_last_move_time()
+        if print_time < self.next_test_time:
+            self.next_test_time = print_time + TEST_TIME
+            return
+        # Raise the bltouch probe and test if probe is raised
+        self.sync_print_time()
+        check_start_time = self.send_cmd('reset', duration=self.pin_move_time)
+        check_end_time = self.send_cmd('touch_mode')
+        self.send_cmd(None)
+        self.verify_state(check_start_time, check_end_time, True,
+                          "verify sensor state")
         # Test was successful
-        self.next_test_time = home_time + TEST_TIME
-        toolhead.reset_print_time(home_time + 2. * MIN_CMD_TIME)
+        self.next_test_time = check_end_time + TEST_TIME
+        self.sync_print_time()
     def home_prepare(self):
         self.test_sensor()
-        toolhead = self.printer.lookup_object('toolhead')
-        print_time = toolhead.get_last_move_time()
-        self.send_cmd(print_time, 'pin_down')
-        self.send_cmd(print_time + self.pin_move_time, 'touch_mode')
-        toolhead.dwell(self.pin_move_time + MIN_CMD_TIME)
+        self.sync_print_time()
+        self.send_cmd('pin_down', duration=self.pin_move_time - MIN_CMD_TIME)
```

Comment:
I don't understand this comment.

Reply:
In short, even if the probe is in an error state, the probing will go ahead. The error is only detected in the finalize step. 

We can also warn about configuration errors here, since at this point, it's very easy to detect if the probe is starting too close to the bed. The probe will simply enter an error state on pin down, which can be cleared after reset. So the most likely cause for that is being too close to the bed, which is a configuration error. Of course it could also be something else, but we can at least give a hint about the configuration being wrong.

Reply:
Yes, as I told in the first paragraph of the initial comment, in practice you are doing the same checks, just in a very backwards way in my eyes.

If I see a function called `test_sensor` I would assume that it performs all the tests it can at that point instead of only doing a partial check, and go on with the homing operation. Then I see a pin down command, which again can be verified at this point, but isn't. Then I see that the homing operation starts, which commands the axis to move. 

At this point we are relying on the `C` side of the firmware to stop the movement before a single step is taken. But if that is broken, there's nothing that prevents us from crashing into the bed.

So it's all about being explicit about what you are doing, and clarifying the code so that other people understand it, without having to know how the whole system is working. But it's also about being more robust to errors, if we had explicit checks here, and they are broken by some regression, we still have the other `C` side checks to stop us from crashing into the bed. The same thing of course applies to the other direction as well. 

The other side of this, is about giving as relevant error messages to the user, or at least in the log files as we can. If we delay the error checking until the finalize step, in the way that you do now, we are lumping these things together as being the same thing:

1. The probe was in an error state even before the probing started
2. The pin failed to move down
3. There was some error with the probe during the probing operation itself
4. The Z-axis failed to move at all, or moved way too slow

Currently all of those errors are reported as "BLTouch failed to deploy", and 2. has the very common cause of being too close to the bed when moving the pin down, so I think that should be part of the error message that the user sees.

I'm sorry for being so critical about this. It has nothing to do with you or that your code is particularly bad. As a matter of fact, you are doing an impressive job, getting this to work this well without owning a BLTouch yourself. And an even more impressive job with the firmware itself. And you are of course allowed to decide what to do with your own code. However, from my point of view, I think that if you are going to present an alternative to my code, you have to improve on it as a whole, not just improve parts of it, and make other parts worse.